File: haskell-etl/src/CsvExtractor.hs
{-# LANGUAGE OverloadedStrings #-}

module CsvExtractor (extractCsv) where

import           WeatherRecord (WeatherRecord(..))
import qualified Data.ByteString.Lazy as BL
import           Data.Csv (HasHeader(..))
import qualified Data.Csv as Csv
import qualified Data.Vector as V
import qualified Data.Text as T
import           Data.Time (parseTimeOrError, defaultTimeLocale)
import           Data.Time.Calendar (Day)
import           Data.Time.LocalTime (LocalTime)

extractCsv :: FilePath -> IO [WeatherRecord]
extractCsv path = do
    csvData <- BL.readFile path
    case Csv.decode HasHeader csvData of
        Left err -> error $ "CSV parse error: " ++ err
        Right rows -> return $ V.toList rows
instance Csv.FromRecord WeatherRecord where
    parseRecord v
        | V.length v == 33 = WeatherRecord
            <$> v Csv..! 0                   -- name
            <*> (parseDay <$> v Csv..! 1)    -- datetime
            <*> v Csv..! 2                   -- tempmax
            <*> v Csv..! 3                   -- tempmin
            <*> v Csv..! 4                   -- temp
            <*> v Csv..! 5                   -- feelslikemax
            <*> v Csv..! 6                   -- feelslikemin
            <*> v Csv..! 7                   -- feelslike
            <*> v Csv..! 8                   -- dew
            <*> v Csv..! 9                   -- humidity
            <*> v Csv..! 10                  -- precip
            <*> v Csv..! 11                  -- precipprob
            <*> v Csv..! 12                  -- precipcover
            <*> (optionalText <$> v Csv..! 13)   -- preciptype
            <*> (optionalDouble <$> v Csv..! 14) -- snow
            <*> (optionalDouble <$> v Csv..! 15) -- snowdepth
            <*> v Csv..! 16                  -- windgust
            <*> v Csv..! 17                  -- windspeed
            <*> v Csv..! 18                  -- winddir
            <*> v Csv..! 19                  -- sealevelpressure
            <*> v Csv..! 20                  -- cloudcover
            <*> v Csv..! 21                  -- visibility
            <*> v Csv..! 22                  -- solarradiation
            <*> v Csv..! 23                  -- solarenergy
            <*> v Csv..! 24                  -- uvindex
            <*> v Csv..! 25                  -- severerisk
            <*> (parseLocalTime <$> v Csv..! 26) -- sunrise
            <*> (parseLocalTime <$> v Csv..! 27) -- sunset
            <*> v Csv..! 28                  -- moonphase
            <*> v Csv..! 29                  -- conditions
            <*> v Csv..! 30                  -- description
            <*> v Csv..! 31                  -- icon
            <*> (optionalText <$> v Csv..! 32)   -- stations
        | otherwise = fail $ "Expected 33 fields, got " ++ show (V.length v)

parseDay :: T.Text -> Day
parseDay = parseTimeOrError True defaultTimeLocale "%Y-%m-%d" . T.unpack

parseLocalTime :: T.Text -> LocalTime
parseLocalTime = parseTimeOrError True defaultTimeLocale "%Y-%m-%dT%H:%M:%S" . T.unpack

optionalText :: T.Text -> Maybe T.Text
optionalText t
    | T.null (T.strip t) = Nothing
    | otherwise          = Just (T.strip t)

optionalDouble :: T.Text -> Maybe Double
optionalDouble t
    | T.null (T.strip t) = Nothing
    | otherwise          = Just (read (T.unpack (T.strip t)))

File: java-etl/src/main/java/br/edu/utfpr/etl/CsvExtractor.java
package br.edu.utfpr.etl;

import com.opencsv.CSVReader;
import com.opencsv.exceptions.CsvException;

import java.io.FileReader;
import java.io.IOException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class CsvExtractor {

    private final String filePath;

    public CsvExtractor(String filePath) {
        this.filePath = filePath;
    }

    public List<WeatherRecord> extract() throws IOException, CsvException {
        List<WeatherRecord> records = new ArrayList<>();

        try (CSVReader reader = new CSVReader(new FileReader(filePath))) {
            List<String[]> rows = reader.readAll();

            // Skip header row
            for (int i = 1; i < rows.size(); i++) {
                String[] cols = rows.get(i);
                records.add(parseRow(cols));
            }
        }

        return records;
    }

    private WeatherRecord parseRow(String[] cols) {
        return new WeatherRecord(
            cols[0],                                        // name
            LocalDate.parse(cols[1]),                       // datetime
            Double.parseDouble(cols[2]),                    // tempmax
            Double.parseDouble(cols[3]),                    // tempmin
            Double.parseDouble(cols[4]),                    // temp
            Double.parseDouble(cols[5]),                    // feelslikemax
            Double.parseDouble(cols[6]),                    // feelslikemin
            Double.parseDouble(cols[7]),                    // feelslike
            Double.parseDouble(cols[8]),                    // dew
            Double.parseDouble(cols[9]),                    // humidity
            Double.parseDouble(cols[10]),                   // precip
            Double.parseDouble(cols[11]),                   // precipprob
            Double.parseDouble(cols[12]),                   // precipcover
            parseOptionalString(cols[13]),                  // preciptype
            parseOptionalDouble(cols[14]),                  // snow
            parseOptionalDouble(cols[15]),                  // snowdepth
            Double.parseDouble(cols[16]),                   // windgust
            Double.parseDouble(cols[17]),                   // windspeed
            Double.parseDouble(cols[18]),                   // winddir
            Double.parseDouble(cols[19]),                   // sealevelpressure
            Double.parseDouble(cols[20]),                   // cloudcover
            Double.parseDouble(cols[21]),                   // visibility
            Double.parseDouble(cols[22]),                   // solarradiation
            Double.parseDouble(cols[23]),                   // solarenergy
            Double.parseDouble(cols[24]),                   // uvindex
            Double.parseDouble(cols[25]),                   // severerisk
            LocalDateTime.parse(cols[26]),                  // sunrise
            LocalDateTime.parse(cols[27]),                  // sunset
            Double.parseDouble(cols[28]),                   // moonphase
            cols[29],                                       // conditions
            cols[30],                                       // description
            cols[31],                                       // icon
            parseOptionalString(cols[32])                   // stations
        );
    }

    private Optional<String> parseOptionalString(String value) {
        if (value == null || value.trim().isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(value.trim());
    }

    private Optional<Double> parseOptionalDouble(String value) {
        if (value == null || value.trim().isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(Double.parseDouble(value.trim()));
    }
}
